editor/src/messages/tool/tool_messages/select_tool.rs

```diff
@@ -626,14 +626,69 @@ impl Fsm for SelectToolFsmState {
+					let transform_vec_round = transform_vec.round();
+					let empty_vec = DVec2 { x: 0.0, y: 0.0 };
 					// TODO: Cache the result of `shallowest_unique_layers` to avoid this heavy computation every frame of movement, see https://github.com/GraphiteEditor/Graphite/pull/481
 					for path in Document::shallowest_unique_layers(tool_data.layers_dragging.iter()) {
-						responses.add_front(GraphOperationMessage::TransformChange {
-							layer: path.to_vec(),
-							transform: DAffine2::from_translation(mouse_delta + closest_move),
-							transform_in: TransformIn::Viewport,
-							skip_rerender: true,
-						});
+						if !(transform_vec_round == empty_vec) && document.grid_enabled {
+							// Get the layer's position in viewport and convert to document space
+							let viewspace = document.document_legacy.generate_transform_relative_to_viewport(path).ok().unwrap_or_default();
+							// Change values of viewspace_port to x: 0.5, y: 0.5 for the grid to be based on pivot
+							let viewspace_pos = viewspace.transform_point2(DVec2 { x: 0.0, y: 0.0 });
+							let doc_pos = document.document_legacy.root.transform.inverse().transform_point2(viewspace_pos);
+
+							// Find the x and y offset on the grid, Given position 1.55px -> 0.55px
+							let grid_offset_x = doc_pos.x.abs() - doc_pos.x.abs().floor();
+							let grid_offset_y = doc_pos.y.abs() - doc_pos.y.abs().floor();
+							let mut adjustment = DVec2 { x: grid_offset_x, y: grid_offset_y };
+
+							// If the original position is negative, flip the grid offset to be used in math to calculate transform
+							if doc_pos.x < 0.0 {
+								adjustment.x = 1.0 - adjustment.x
+							}
+							if doc_pos.y < 0.0 {
+								adjustment.y = 1.0 - adjustment.y
+							}
+
+							// Check if the x and y offsets are on a whole number
+							let scaling_factor = 100.0;
+							let rounded_grid_x = (grid_offset_x * scaling_factor).round() / scaling_factor;
+							let rounded_grid_y = (grid_offset_y * scaling_factor).round() / scaling_factor;
+							let x_aligned = rounded_grid_x % 1.0 == 0.0;
+							let y_aligned = rounded_grid_y % 1.0 == 0.0;
+
+							// If the x or y positions are off the grid, based on direction of mouse movement calculate the new transform that positioned on the grid
+							if !x_aligned && transform_vec_round.y == 0.0 {
+								if transform_vec.x > 0.0 {
+									new_transform_vec.x = new_transform_vec.x.abs() * (1.0 - adjustment.x);
+								} else if transform_vec.x < 0.0 {
+									new_transform_vec.x = new_transform_vec.x.abs() * adjustment.x * -1.0;
+								}
+							}
+							if !y_aligned && transform_vec_round.x == 0.0 {
+								if transform_vec.y > 0.0 {
+									new_transform_vec.y = new_transform_vec.y.abs() * (1.0 - adjustment.y);
+								} else if transform_vec.y < 0.0 {
+									new_transform_vec.y = new_transform_vec.y.abs() * adjustment.y * -1.0;
+								}
+							}
+
+							responses.add_front(GraphOperationMessage::TransformChange {
```

Comment:
this responses.add block is repeated twice, could be cleaner to just change "transform_vec" and have the responses.add outside the if and else

Reply:
good idea. Just made the changes